NickNackGus/Remote Player Manager

NickNackGus opened the pull request.

Ascynx committed:
RemotePlayerManager Port

Added the ability to register plugin data on the remote player messages.

Added event GatherRemotePlayerDataEvent to gather the plugin data to be added on the remote player messages.

Added events RemotePlayerLoadedEvent & RemotePlayerUnloadedEvent to get an easier way to listen to changes in the full online player list.

Ascynx committed:
PR Reviews part 1

Added back vanished default variables

Switched to using Optionals for values where the manager could be null.

A shard switch will now only be 1 load/unload event, even when a race condition would happen.

Ascynx committed:
Moved stuff around to make it follow how it was formatted in the network-chat counterpart.

Ascynx committed:
Switched to fully static (and so no more optional)

Ascynx committed:
Forgot about 2-3 things

NickNackGus committed:
Code review changes

Ascynx committed:
nullables go brrrr

NickNackGus committed:
Fixed missing player name in log message

NickNackGus committed:
Fixed unused imports

NickNackGus committed:
Merge branch 'ascynx/rpm-port' of github.com:Ascynx/monumenta-network-relay into ascynx/rpm-port

NickNackGus committed:
Marked many API-accessible methods as protected again

Ascynx committed:
Abstracted RemotePlayerManager

(surely I did it correctly)

Ascynx committed:
override

Ascynx committed:
forgot about nullables

Ascynx committed:
reviewdoggo did not like it

Ascynx committed:
maybe this will do

Ascynx committed:
tab indents this time around

Ascynx committed:
Moved remote player attributes around

Moved vanished to RemotePlayerPaper

Moved Plugin data to RemotePlayerAbstraction

Added Proxy name to RemotePlayerBungee (should be moved to the Bungee implementation of RemotePlayerManager once made)

Ascynx committed:
oops forgot about that

Ascynx committed:
does it want the abstract before static?

NickNackGus committed:
Updated RemotePlayer data types to be loadable from anywhere; RemotePlayerPaper can be created locally on Paper

NickNackGus committed:
Fixed missing argument

U5B committed:
usb: expose a way to get all player uuids

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
usb: making abstract contain common data 1

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
more abstracted stuff 2

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
I probably did something wrong
:suffer:

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add more null checks
removed unneeded methods
only update null proxy, shard and world values

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
create Bungee portion
run gatherPluginData outside of RemotePlayerPaper

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add new API

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add even more API

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
only broadcast local data

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
only update values that aren't local
add missing load event to bungee

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
update plugin data on shard transfer
add /remoteplayerinfo command for debug purposes
format RemotePlayerAbstraction.toString

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
Merge branch 'master' into upstream/usb-rpm

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
fix channel name
found better way of removing from proxy/shard lists

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

NickNackGus committed:
Fixed RemotePlayerAbstraction types, infrastructure needs updating

NickNackGus committed:
Renamed server type bungee -> proxy

NickNackGus committed:
WIP logic changes

NickNackGus commented:
I *think* the data structure should be under control now, though the supporting logic hasn't been updated to match. I'm not sure if I'll be able to get back to this until the end of the week, we'll see.

NickNackGus committed:
Check destination type

NickNackGus committed:
List server types player is on

NickNackGus committed:
Theoretically updated RPMAbstraction sufficiently, implementations not updated

NickNackGus committed:
Fixed isHidden() check

NickNackGus committed:
Corrected use of RPMAbstraction in API

NickNackGus committed:
Made isHidden optional

NickNackGus committed:
WIP changes, Bungee/Paper implementations need work

U5B committed:
wip: bungee/paper immplimentation
- missing commands/events

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add json formatted toString
force logging to be finest

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
update to Java 17 and Paper 1.19.4
- fix isHidden
- add id to IllegalStateException (probably should be removed)
- targetShard is fricked by bungee stupidity until we migrate to Velocity (soon:tm:)
- fed the reviewdog a tiny bit

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add whereis and remoteplayerapi command

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

NickNackGus committed:
autofix imports and indentation

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/WhereIsCommand.java:
```diff
@@ -0,0 +1,36 @@
+			.withArguments(new EntitySelectorArgument.OnePlayer("player"))
+			.executes((sender, args) -> {
+				Player player = (Player) args[0];
+				// TODO: REMOVE THIS COMMAND, THIS IS FOR DEBUGGING DEV/MOD only - usb
```
This is a command requested by the moderator team. Also note that using an entity selector argument requires that you already know where the player is and run the command from their shard, which is counterproductive. This needs to use a TextArgument with suggestions for the names of visible player names - or also include hidden player names if the CommandSender has the permission to see them, but I never got around to figuring that out. An operator check is not sufficient to verify this due to the build shard.

U5B replied:
TODO comment was copied from a debugging command

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerAPICommand.java:
```diff
@@ -0,0 +1,35 @@
+					} else {
+						player = sender;
+					}
+					// TODO: REMOVE THIS COMMAND, THIS IS FOR DEBUGGING DEV/MOD only - usb
```
Since this is likely to be useful for development/debugging of anything using this API, it would be good to keep this, just restricted to people with the permission. Same suggestion for the argument type as the `/whereis` command.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerAbstraction.java:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	protected void broadcast() {
+		JsonObject playerData = this.toJson();
```
`this.` is typically not required and so we usually skip it; was it using the abstraction's version of `toJson()` instead of the specific type's version? If so, a comment may prove helpful for anyone looking at this later.

U5B replied:
It isn't I can revert it
It was because I was investigating the mTargetShard value being null but it was just bungee being dumb.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,185 @@
+	}
+
+	protected static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline) {
+		String targetShard = ""; // TODO: add a way to get the shard name. IT IS MUCH EASIER IN VELOCITY - usb
```
It's better to make this nullable so it's more obvious the value is unknown/unset - unless you think there's a way to get this value from bungee before switching to velocity, in which case we should either implement it or switch to Velocity before this value gets used.

NickNackGus replied:
The concern being that if we were to use this value as a tie-breaker when transferring players from one shard to another, we don't want it to claim they're on neither in an unintuitive way.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,185 @@
+	protected RemotePlayerAbstraction remotePlayerChange(JsonObject data) {
+		RemotePlayerAbstraction player = RemotePlayerAbstraction.from(data);
+		if (player == null) {
+			// something really bad happened
```
Well that shouldn't happen - best to put a log message here so we at least know if it does.

NickNackGus commented:
I didn't spot any other issues so far - though this will need thorough testing to be sure. Fun with systems that *must* be correct, plus or minus some short delay with eventual consistency.

U5B committed:
resolve comments from Nick
- add javadoc comments
- remove some TODO

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
fail on reviewdog error

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
usb: add events to paper/bungee
- get mTargetShard to return a value
- whereis command has tabcomplete for visible players and hides vanished players
- added 1 tick delay to paper quit event due to race refresh

RemotePlayerLoadedEvent: fires when a player is loaded locally or remotely and they weren't online previously
RemotePlayerUnloadedEvent: fires when a player is unloaded locally or remotely and they were online previously
RemotePlayerUpdatedEvent: fires if the player data hs changed and the two above haven't fired
Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
fix counter correction refresh not working
-  add isSimilar to RemotePlayerAbstraction that ignores plugindata differences
- feed the extremely hungry reviewdog 1

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
feed the reviewdog 2

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
feed the reviewdog 3

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
fix checkstyle not excluding eventhandler for bungee

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerAbstraction.java:
```diff
@@ -0,0 +1,155 @@
+
+	/** Determine the appropriate remote player data type to use */
+	public static @Nullable RemotePlayerAbstraction from(JsonObject remoteData) {
+		if (remoteData == null) {
```
remoteData is not nullable - this should never be allowed to happen, which means this method does not need to return nullable either.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,216 @@
+	}
+
+	static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline) {
+		@Nullable String targetShard = player.getServer() != null ? player.getServer().getInfo().getName() : ""; // TODO: add a way to get the shard name. IT IS MUCH EASIER IN VELOCITY - usb
```
Doesn't this already get the shard name? Shouldn't matter too much anyways if we're switching to Velocity.

U5B replied:
This does get the shard name, but the events that fire before the MonumentaReconnectHandler don't have this information populated. That includes `PostLoginEvent`.
However `ServerSwitchEvent` has this data, so we use that to populate `mTargetShard`
I can probably remove the TODO however

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,216 @@
+		if (isRemote && serverType.equals(RemotePlayerBungee.SERVER_TYPE) && localPlayer != null && localPlayer.isConnected()) {
+			// Player logged off on remote shard, but is locally online.
+			// This can happen if the remote shard was not notified the player logged in here in time.
+			MMLog.warning(() -> "Detected race condition, triggering refresh on " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
```
This is not severe enough to be a warning - it can happen fairly often as players switch shards and should resolve itself fast enough to be a non-issue.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,216 @@
+			return false;
+		}
+
+		if (!player.mIsOnline && (oldPlayer == null || oldPlayer.mIsOnline)) {
```
If oldPlayer is null, they are theoretically already offline (or otherwise were never registered with this shard because it was starting up). The logic might need tweaking here.

NickNackGus replied:
Related note, it should be possible to handle the logic of whether or not an event needs to be sent in the abstraction layer so that part of the logic isn't repeated, with methods to call the appropriate event in the implementations.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,216 @@
+			MMLog.info(() -> "Updated player: " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
+			return true;
+		} else {
+			MMLog.warning(() -> "Ignored player: " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
```
I don't think ignoring data should be a warning as a general rule - either there's nothing to update, and it can be `fine` or lower priority, or something went wrong trying to understand remote data, which should be a warning with more info.

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,216 @@
+		updatePlayerLocal(player, true);
+	}
+
+	boolean updatePlayerLocal(RemotePlayerAbstraction player, boolean isRemote) {
```
This could use a comment block explaining the return value; it appears that it returns false if the local value is correct and doesn't need updating. Also consider the name `updateLocalCache` or similar, which should avoid confusion with why updating something local needs a flag for remote data. (Though that might be on me for how I named things the first time around.)

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java:
```diff
@@ -0,0 +1,216 @@
+			localPlayer.broadcast();
+		}
+	}
+
```
There's a commented out method that was removed from here since commit `3227a0243970f41b025581204a5e67b87bf9cc9e` that would have handled listening for players switching shards, which would update the `proxy` server data about which shard a player is supposedly on.

If the code elsewhere in this file can detect which shard the player is on, listening for that may still prove helpful. It'd also be ideal if the `minecraft` (Paper) server type could check for the player's shard from the `proxy` server type before marking them as having changed shards - though that needs checks to ensure there's a valid value to compare against.

U5B replied:
The commented out event in question `ServerSwitchEvent` is still here: just that I moved it without the comment
whoops

NickNackGus commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerPaper.java:
```diff
@@ -0,0 +1,247 @@
+package com.playmonumenta.networkrelay;
```
...I'm not checking this file thoroughly for now, I'm just going to assume it has similar issues to the Bungee implementation.

NickNackGus commented:
Some other stuff that we missed:
- There's one API currently, and it only works from Paper since it requires CommandAPI
- There's no way for a command or API to request a refresh on a local player should their data change (mainly NetworkChat name formatting for now)

U5B committed:
merge flowey's ssh changes

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
move updatePlayerLocal to updateLocalPlayer in abstraction
- rewrite log messages to occur after the event
- added methods for abstraction to call for bungee/paper
- RemotePlayerAbstraction.from no longer returns null
- added comment about player.getServer behavior on bungee
- downgraded race condition warning to info
- downgraded ignored player warning to fine

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add ability to refresh player from api

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add way to set plugin data

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
add temp whereisbungee command
move argumentssuggestions to commands

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
feed reviewdoggo

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
fix ALL_PLAYER_ONLINE_SUGGESTIONS returning only visible players

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

U5B committed:
only show visible players

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

Nobody314159 committed:
merge origin/master, update commands

NickNackGus committed:
Renamed data-only Minecraft/proxy classes

NickNackGus committed:
Standardized on isPlayerVisible due to missing logical nots

U5B committed:
don't bundle adventure

Signed-off-by: U5B <56985400+U5B@users.noreply.github.com>

NickNackGus committed:
Added API for checking destination types

NickNackGus committed:
Go away useless deprecation warning

NickNackGus committed:
Unused imports

NickNackGus committed:
/whereis uses friendly string

NickNackGus committed:
Moved unregistering server before unregistering server type

NickNackGus committed:
Further /whereis cleanup

NickNackGus commented:
This PR can be deployed to the **build** server after this week's update in order to test it with https://github.com/TeamMonumenta/monumenta-network-chat/pull/67. If there are no warnings after one week, it can ship to play. If there are no warnings a week on play, it is safe to use this API for any player data that almost never updates between a player joining a shard and leaving it.

Nobody314159 commented on src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerPaper.java:
```diff
@@ -0,0 +1,253 @@
+
+		if (command.startsWith("/pv ")
+			|| "/pv".equals(command)
+			|| command.contains("vanish")) {
```
These checks seem a bit weird - trying to be careful with pv but vanish could appear anywhere. Fortunately doesn't matter much since the worst case scenario is refreshing slightly more than necessary. Maybe could get the substring of the command ending until the first space and check that that is exactly "/pv" or "vanish"?

NickNackGus replied:
Yeah, this is probably overly cautious, though it's the same check as in the chat plugin now. There just isn't really a good way to check if a player's vanished state changes without using `/pv` with optional arguments, or hoping that if another vanish command is in use, it contains `vanish`. Though I should probably at least include a leading `/` in all cases so it can't trigger on general chat.

Nobody314159 reviewed (approved):
I can't claim to understand what's happening in the big picture but every piece of code looks good to me

NickNackGus committed:
IntelliJ fixes

NickNackGus commented:
Let the beta testing begin!
